projects/plugins/crm/src/automation/class-base-step.php

```diff
@@ -188,14 +188,29 @@ public function validate( Data_Type $data ): void {
 		}
 	}
 
+	/**
+	 * Validate and execute the step.
+	 *
+	 * @since $$next-version$$
+	 *
+	 * @param Data_Type $data Data type passed.
+	 * @return void
+	 *
+	 * @throws Data_Type_Exception If the data type passed is not expected.
+	 */
+	public function validate_and_execute( Data_Type $data ): void {
+		$this->validate( $data );
+		$this->execute( $data );
+	}
+
 	/**
 	 * Execute the step.
 	 *
 	 * @since $$next-version$$
 	 *
 	 * @param Data_Type $data Data type passed from the trigger.
 	 */
-	abstract public function execute( Data_Type $data );
+	abstract protected function execute( Data_Type $data );
```

Comment:
I made it `protected`, to force using the `validate_and_execute` method instead of.